apollo-relay/src/lib.rs

```diff
@@ -0,0 +1,532 @@
+                .body(compressed_content)
+                .header("X-Api-Key", key)
+                .header("Content-Encoding", "gzip")
+                .header("Content-Type", "application/protobuf")
+                .header("Accept", "application/json")
+                .send()
+                .await
+                .map_err(|e| Status::failed_precondition(e.to_string()))?;
+            tracing::debug!("result: {:?}", res);
+            let data = res
+                .text()
+                .await
+                .map_err(|e| Status::internal(e.to_string()))?;
+            tracing::debug!("text: {:?}", data);
+            /*
+            let ar: ApolloResponse = res
+                .json()
+                .await
+                .map_err(|e| Status::internal(e.to_string()))?;
+            tracing::debug!("json: {:?}", ar);
+            */
+            let response = ReporterResponse {
+                message: "Report accepted".to_string(),
+            };
+            Ok(Response::new(response))
+        }
+    }
+
+    #[tonic::async_trait]
+    impl Reporter for ReportRelay {
+        async fn add_stats(
+            &self,
+            request: Request<ReporterStats>,
+        ) -> Result<Response<ReporterResponse>, Status> {
+            tracing::debug!("received request: {:?}", request);
+            let msg = request.into_inner();
+            self.add_stats_or_trace(StatsOrTrace::Stats(msg)).await
+        }
+
+        async fn add_trace(
+            &self,
+            request: Request<ReporterTrace>,
+        ) -> Result<Response<ReporterResponse>, Status> {
+            tracing::debug!("received request: {:?}", request);
+            let msg = request.into_inner();
+            self.add_stats_or_trace(StatsOrTrace::Trace(msg)).await
+        }
+    }
+
+    impl ReportRelay {
+        async fn add_stats_or_trace(
+            &self,
+            record: StatsOrTrace,
+        ) -> Result<Response<ReporterResponse>, Status> {
+            let response = ReporterResponse {
+                message: "Report accepted".to_string(),
+            };
+            let mut tpq = self.tpq.lock().await;
+            let graph_map = tpq
+                .entry(record.graph().unwrap())
+                .or_insert_with(HashMap::new);
+            let entry = graph_map
+                .entry(record.key())
+                .or_insert_with(|| record.get_traces_and_stats());
+            match record {
+                StatsOrTrace::Stats(mut s) => {
+                    entry.stats_with_context.push(s.stats.take().unwrap())
+                }
+                StatsOrTrace::Trace(mut t) => entry.trace.push(t.trace.take().unwrap()),
+            }
+
+            // Trigger a dispatch if we have "too much" data
+            let mut total = 0;
+            for (_graph, entry) in tpq.iter() {
+                total += entry.len();
+            }
+
+            if total > 10 {
+                let mut backoff = 0;
```

Comment:
not a rust expert but feels like this could be itself a Duration that has other Durations added to it rather than doing integer math and only converting to Duration later? means I don't have to be confused about what the unit of "500" later is

Reply:
Yup. I'll clean that up.